internal/rpc/digest_md5.go

```diff
@@ -0,0 +1,473 @@
+//
+//   If "qop" is "auth-int" or "auth-conf":
+//
+//     A2 = { "AUTHENTICATE:", digest-uri-value, ":00000000000000000000000000000000" }
+//
+//   Where:
+//
+//     { a, b, ... } is the concatenation of the octet strings a, b, ...
+//
+//     H(s) is the 16 octet MD5 Hash [RFC1321] of the octet string s
+//
+//     KD(k, s) is H({k, ":", s}), i.e., the 16 octet hash of the string k, a colon, and the string s
+//
+//     HEX(n) is the representation of the 16 octet MD5 hash n as a string of 32 hex digits (with alphabetic characters
+//     in lower case)
+func compute(a1, a2, nonce, cnonce, ncvalue, qop string) string {
+	x := hexfn(h(a1))
+	y := strings.Join([]string{nonce, ncvalue, cnonce, qop, hexfn(h(a2))}, ":")
+	return hexfn(kd(x, y))
+}
+
+func h(s string) [md5.Size]byte {
+	return md5.Sum([]byte(s))
+}
+
+func kd(k, s string) [md5.Size]byte {
+	return h(strings.Join([]string{k, s}, ":"))
+}
+
+func hexfn(data [md5.Size]byte) string {
+	return hex.EncodeToString(data[:])
+}
+
+func a1(username, realm, password, nonce, cnonce string) string {
+	x := h(strings.Join([]string{username, realm, password}, ":"))
+	return strings.Join([]string{string(x[:]), nonce, cnonce}, ":")
+}
+
+func a2(digestURI, qop string, initial bool) string {
+	var a2 string
+	// when computing the initial response digest, we use the "AUTHENTICATE" piece
+	// but when validating the server response-auth, the "AUTHENTICATE:" string is
+	// left out in taht calculation to confirm the digest-response
+	if initial {
+		a2 = strings.Join([]string{"AUTHENTICATE", digestURI}, ":")
+	} else {
+		a2 = ":" + digestURI
+	}
+	if qop == qopPrivacy || qop == qopIntegrity {
+		a2 = a2 + ":00000000000000000000000000000000"
+	}
+	return a2
+}
+
+// If the server offered qop=auth-int and we replied with qop=auth-int then subsequent
+// messages need to be integrity protected. The base session key is H(A1) as defined above
+// this function will generate the pair of integrity keys for both client to server (kic)
+// and server to client (kis) as defined in the RFC with the specified magic constants
+func generateIntegrityKeyPair(a1 string) ([md5.Size]byte, [md5.Size]byte) {
+	clientIntMagicStr := []byte("Digest session key to client-to-server signing key magic constant")
+	serverIntMagicStr := []byte("Digest session key to server-to-client signing key magic constant")
+
+	sum := h(a1)
+	kic := md5.Sum(append(sum[:], clientIntMagicStr...))
+	kis := md5.Sum(append(sum[:], serverIntMagicStr...))
+
+	return kic, kis
+}
+
+// If message integrity is negotiated, there will be a MAC block for each message appended
+// to the message. The MAC block is 16 bytes, the first 10 bytes of the HMAC-MD5 of the message
+// plus a 2 byte message type number and a 4 byte sequence number. This function provides the
+// HMAC as defined in the RFC as HMAC(ki, {seqnum, msg})[0..9]
+func getHMAC(key, seq, msg []byte) []byte {
+	data := append(seq, msg...)
+
+	mac := hmac.New(md5.New, key)
```

Comment:
This is unfortunately hiding an allocation: https://golang.org/src/crypto/hmac/hmac.go?s=1932:1982#L60

We can reuse one and save it on the struct.